packages/hardhat-verify/src/internal/etherscan.ts

```diff
@@ -0,0 +1,251 @@
+      const responseText = await response.body.text();
+      throw new ContractVerificationInvalidStatusCodeError(
+        this._apiUrl,
+        response.statusCode,
+        responseText
+      );
+    }
+
+    const etherscanResponse = new EtherscanResponse(await response.body.json());
+
+    if (etherscanResponse.isBytecodeMissingInNetworkError()) {
+      throw new ContractVerificationMissingBytecodeError(
+        this._apiUrl,
+        contractAddress
+      );
+    }
+
+    if (!etherscanResponse.isOk()) {
+      throw new HardhatVerifyError(etherscanResponse.message);
+    }
+
+    return etherscanResponse;
+  }
+
+  // https://docs.etherscan.io/api-endpoints/contracts#check-source-code-verification-submission-status
+  public async getVerificationStatus(guid: string): Promise<EtherscanResponse> {
+    const parameters = new URLSearchParams({
+      apikey: this._apiKey,
+      module: "contract",
+      action: "checkverifystatus",
+      guid,
+    });
+    const url = new URL(this._apiUrl);
+    url.search = parameters.toString();
+
+    let response;
+    try {
+      response = await sendGetRequest(url);
+    } catch (error: any) {
+      throw new ContractStatusPollingError(url.toString(), error);
+    }
+
+    if (!(response.statusCode >= 200 && response.statusCode <= 299)) {
+      // This could be always interpreted as JSON if there were any such guarantee in the Etherscan API.
+      const responseText = await response.body.text();
+
+      throw new ContractStatusPollingInvalidStatusCodeError(
+        response.statusCode,
+        responseText
+      );
+    }
+
+    const etherscanResponse = new EtherscanResponse(await response.body.json());
+
+    if (etherscanResponse.isPending()) {
+      await sleep(VERIFICATION_STATUS_POLLING_TIME);
+
+      return this.getVerificationStatus(guid);
+    }
+
+    if (etherscanResponse.isFailure()) {
+      return etherscanResponse;
+    }
+
+    if (!etherscanResponse.isOk()) {
+      throw new ContractStatusPollingResponseNotOkError(
+        etherscanResponse.message
+      );
+    }
+
+    return etherscanResponse;
+  }
+
+  public getContractUrl(address: string) {
+    return `${this._browserUrl}/address/${address}#code`;
+  }
+}
+
+class EtherscanResponse {
+  public readonly status: number;
+
+  public readonly message: string;
+
+  constructor(response: any) {
+    this.status = parseInt(response.status, 10);
+    this.message = response.result;
+  }
+
+  public isPending() {
+    return this.message === "Pending in queue";
+  }
+
+  public isFailure() {
+    return this.message === "Fail - Unable to verify";
+  }
+
+  public isSuccess() {
+    return this.message === "Pass - Verified";
+  }
+
+  public isBytecodeMissingInNetworkError() {
+    return this.message.startsWith("Unable to locate ContractCode at");
+  }
+
+  public isOk() {
+    return this.status === 1;
+  }
+}
+
+function resolveApiKey(apiKey: ApiKey | undefined, network: string) {
```

Comment:
Can you expose this function as well (e.g. make this a `public static` function in the `Etherscan` class, or provide a way to get the `_apiKey` field from the `Etherscan` class with the resolved value)?

We would like to use the resolved API key to make some separate calls to the Etherscan API, outside of the Etherscan class.

Reply:
Would it be enough if we make `_apiKey` public?

Reply:
Yes, that would work.

Reply:
We had the same thought 😄

Reply:
By the way, I'm going to split this PR into two to avoid blocking you, since the Sourcify part will likely require a major release. You can continue reviewing [here](https://github.com/NomicFoundation/hardhat/pull/4198).